android/java/org/chromium/chrome/browser/playlist/kotlin/playback_service/VideoPlaybackService.kt

```diff
@@ -0,0 +1,297 @@
+        var currentPlaylistId: String = ""
+
+        private val mutableCurrentPlayingItem = MutableLiveData<String>()
+        val currentPlayingItem: LiveData<String>
+            get() = mutableCurrentPlayingItem
+
+        private fun setCurrentPlayingItem(currentPlayingItemId: String) {
+            mutableCurrentPlayingItem.value = currentPlayingItemId
+        }
+
+        private var mediaItemsInPlayer: ArrayList<MediaItem> = ArrayList()
+
+        fun addNewPlaylistItemModel(newPlaylistItem: PlaylistItem) {
+            if (ConstantUtils.DEFAULT_PLAYLIST == currentPlaylistId) {
+                val mediaItem =
+                    MediaItemUtil.buildMediaItem(
+                        newPlaylistItem,
+                        ConstantUtils.DEFAULT_PLAYLIST,
+                        ContextUtils.getApplicationContext()
+                            .resources
+                            .getString(
+                                R.string.playlist_play_later
+                            ),
+                    )
+                mediaItemsInPlayer.add(mediaItem)
+                mPlayer.addMediaItem(mediaItem)
+            }
+        }
+
+        fun removePlaylistItemModel(playlistItemId: String) {
+            mediaItemsInPlayer.forEachIndexed { index, mediaItem ->
+                if (mediaItem.mediaId == playlistItemId && mediaItem.mediaId != currentPlaylistId) {
+                    mPlayer.removeMediaItem(index)
+                    mediaItemsInPlayer.removeAt(index)
+                }
+            }
+        }
+
+        fun reorderPlaylistItemModel(playlistItemList: List<PlaylistItem>) {
+            mediaItemsInPlayer.forEachIndexed { oldIndex, mediaItem ->
+                val newIndex = playlistItemList.indexOfFirst { it.id == mediaItem.mediaId }
+                mPlayer.moveMediaItem(oldIndex, newIndex)
+            }
+        }
+
+        fun getMediaItemIndex(playlistItemId: String): Int {
+            return mediaItemsInPlayer.indexOfFirst { it.mediaId == playlistItemId }
+        }
+    }
+
+    override fun onCreate() {
+        super.onCreate()
+        initPlaylistService()
+        initializeSessionAndPlayer()
+        lastSavedPositionTimer()
+    }
+
+    override fun onConnectionError(mojoException: MojoException) {
+        mPlaylistService?.close()
+        mPlaylistService = null
+        if (
+            ChromeFeatureList.isEnabled(BraveFeatureList.BRAVE_PLAYLIST) &&
+                ChromeSharedPreferences.getInstance()
+                    .readBoolean(BravePreferenceKeys.PREF_ENABLE_PLAYLIST, true)
+        ) {
+            initPlaylistService()
+        }
+    }
+
+    private fun initPlaylistService() {
+        if (mPlaylistService != null) {
+            mPlaylistService = null
+        }
+        mPlaylistService =
+            PlaylistServiceFactoryAndroid.getInstance()
+                .getPlaylistService(ProfileManager.getLastUsedRegularProfile(), this)
```

Comment:
the profile could be not initialized at that point, there is a crash in the current nightly:
```
Exception java.lang.RuntimeException:
  at android.app.ActivityThread.handleCreateService (ActivityThread.java:4732)
  at android.app.ActivityThread.-$$Nest$mhandleCreateService
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:2288)
  at android.os.Handler.dispatchMessage (Handler.java:106)
  at android.os.Looper.loopOnce (Looper.java:201)
  at android.os.Looper.loop (Looper.java:288)
  at android.app.ActivityThread.main (ActivityThread.java:8194)
  at java.lang.reflect.Method.invoke
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:703)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:911)
Caused by java.lang.IllegalStateException: Browser hasn't finished initialization yet!
  at org.chromium.chrome.browser.profiles.ProfileManager.getLastUsedRegularProfile (ProfileManager.java:98)
  at org.chromium.chrome.browser.playlist.hls_content.HlsServiceImpl.initPlaylistService (HlsServiceImpl.java:229)
  at org.chromium.chrome.browser.playlist.hls_content.HlsServiceImpl.onCreate (HlsServiceImpl.java:61)
  at org.chromium.chrome.browser.base.SplitCompatService.onCreate (SplitCompatService.java:37)
  at android.app.ActivityThread.handleCreateService (ActivityThread.java:4719)
```
We should wait till profile is initialized like here https://github.com/chromium/chromium/blob/main/chrome/android/java/src/org/chromium/chrome/browser/download/DownloadManagerService.java#L1014

Reply:
ah in that case, we need to add profile observer and init service once it's fully initialized.